Add simple integration test for openssl_certificate

MarkusTeufelberger opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Add a simple integration test for the openssl_certificate module.
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
* crypto/openssl_certificate
* test/integration
##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
2.4
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
Not a lot of functionality yet, but it is a start. The design is modeled after the openssl_csr integration test.
<!--- Paste verbatim command output below, e.g. before and after your change -->

Commit:
Add simple integration test for openssl_certificate

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/36244

Spredzy commented on test/integration/targets/openssl_certificate/tests/validate.yml:
```diff
@@ -0,0 +1,23 @@
+- name: Validate certificate (assert)
+  assert:
+    that:
+      # Certificate is selfsigned: subject == issuer)
```
IIUC [RFC3280](https://tools.ietf.org/html/rfc3280) (section6.1)

> A certificate is self-issued if the DNs that appear in the subject
>   and issuer fields are identical and are not empty

So what I understand is that this is true for 'self-issued' certificate, but not a requirement for 'self-signed' certificate.

Hence in this scenario, the ouput of `openssl x509 -noout -issuer -in {{ output_dir }}/cert.pem -nameopt oneline,-space_eq` is `issuer=` which I think is valid.

However, we could add an option to the self-signed backend to eventually allow one to specify the issuer via [set_issuer()](https://pyopenssl.org/en/stable/api/crypto.html#OpenSSL.crypto.X509.set_issuer)

What are your thoughts on this ?

MarkusTeufelberger replied:
That's definitely an option that is fine by me, it seemed just like an easy assertion to initilly try out. I'm more concerned about the CI failure when checking the private key in the module itself.

Spredzy replied:
I don't unserstand: 

> I'm more concerned about the CI failure when checking the private key in the module itself

The following assertion works:

```
 - name: Validate certificate (assert)      
    assert:                                  
        that:                                                                          
          - cert_modulus.stdout == privatekey_modulus.stdout 
```

What are you exactly refering to ?

MarkusTeufelberger replied:
https://app.shippable.com/github/ansible/ansible/runs/36244/21/console for example

Spredzy replied:
@MarkusTeufelberger ok, I found the root cause issue. I'll provide a patch as soon as I find an elegant solution for it.

Spredzy replied:
@MarkusTeufelberger could you integrate this commit https://github.com/Spredzy/ansible/commit/a25db3d61f69a6d5f82ce67d7ed2abff199bb539 in your PR (if you agree with its content - else let me know if there is anything you'd like to see modified)?

This will fix the issue you've spotted. Once this PR merged, we will need to backport it for RC2 since you've found an actual issue with the code. Thanks again for that.

MarkusTeufelberger replied:
Are dictionary parameters (subject, issuer) internally parsed into tuples? Otherwise it might be necessary to add a `elif isinstance(attr, dict):` line there, right?

Spredzy replied:
Yep, they are internally parsed into tuple. Hence the tuple test. None of the current input are parsed as dict. so that why the test isn't present.

MarkusTeufelberger replied:
Great, thanks! :-)

Spredzy replied:
Ideally, you should put commit a4026d0 before c66ccce in your branch so tests are always ok no matter on which commit it is run. Also for now I think we can drop the subject.stdout == issuer.stdout until we add this feature in the code.

Spredzy committed:
openssl_certificate: Fix parameter assertion in Python3

Parameter assertion in Python3 is broken. pyOpenSSL get_X() functions
returns b'' type string and tries to compare it with '' string, leading
to failure.

The error mentionned above has been fixed by sanitizing the inputs from
a user to the assert only backend.

Also, this error was hidden by the fact that the improper check method
was called in the generate() functions.

Commit:
remove subject == issuer assertion

Spredzy commented:
@MarkusTeufelberger `add_extensions()` appeared in `0.15.x` so you need to condition the test to run only on the nodes that meets the requirements. Example is available here https://github.com/ansible/ansible/blob/devel/test/integration/targets/openssl_publickey/tasks/main.yml#L53

I will on my side, ensure it fails properly, but this shouldn't be a blocker for this PR.

Commit:
run integration tests only on supported hosts

Spredzy commented on test/integration/targets/openssl_certificate/tasks/main.yml:
```diff
@@ -1,30 +1,33 @@
-- import_tasks: ../tests/validate.yml
+  - import_tasks: ../tests/validate.yml
+
+when: pyopenssl_version.stdout|version_compare('16.0.0', '>=')
```
Missing two spaces and the version should be 0.15.0 and up, not 16.0.0

Commit:
change min supported version to 0.15.x

MarkusTeufelberger commented:
Shouldn't this also be a problem in openssl_csr and lots of other modules as well?!

MarkusTeufelberger commented:
`openssl_csr` uses a lot of `.encode('ascii')` when using pyopenssl...

Commit:
Add test for more CSR fields

Spredzy committed:
openssl_{csr, certificate}: Fail if pyOpenSSL <= 0.15

Previous 0.13 pyOpenSSL was a C-binding, and required the parameter
passed to add_extention to be in ASN.1. This has changed with the move
to 0.14 and it is now all pythong and string based.

Previous the 0.15 release, the `get_extensions()` method didn't exist,
since the modules rely heavily on it we ensure pyOpenSSL version is at
last 0.15.0.

Spredzy commented:
@MarkusTeufelberger 

So the issue:

  * RHEL/CentOS ships with `pyOpenSSL==0.13.1` which were C-bindings and not a pure python implementations, and it somehow expects an `ASN.1` formatted string when using `X509Extension()`. Pure python extensions came with `pyOpenSSL==0.14`
  * `get_extensions()` exists since `pyOpenSSL==0.15`, since both `openssl_csr` and `openssl_certificate` rely on it, it makes sense to make it mandatory.

So if we look at the Operating Systems tested by the CI system, they either ship with `pyOpenSSL==0.13.x` or `pyOpenSSL=16.x.x` as far as I can tell. So summing up all this, it made sense to me, to make `pyOpenSSL==0.15` mandatory for both `openssl_csr` and `openssl_certificate`. Here is a commit doing it:
https://github.com/Spredzy/ansible/commit/4c8317c8f02ac9978f5e55f508a832e8cb0068b0

Let me know if you agree/disagree with it.

Note: on the version compare the version should be `0.15` and not `15.0.0`

MarkusTeufelberger commented:
Yes, it makes sense to enforce this a bit more.

Currently I'm still trying to debug an issue that came up when running with python3 where the output from pyOpenSSL is a byte-string and it gets compared to an "actual" string in the subject or issuer parameters. Something that shouldn't happen any more...

Commit:
also convert dict members to bytes

Commit:
fix version_compare

MarkusTeufelberger commented:
Seems like on python3, dict parameters can also be an instance of `dict` (https://github.com/ansible/ansible/pull/29038/commits/29f0a11b23214bb3afb278ecdb0072d487e0874e), at least our errors went away after this change.

Next up: Certificates with more than one organization. Apparently a common use case for Kubernetes: https://kubernetes.io/docs/admin/authentication/#x509-client-certs:

> As of Kubernetes 1.4, client certificates can also indicate a user’s group memberships using the certificate’s organization fields. To include multiple group memberships for a user, include multiple organization fields in the certificate.

Also an upcoming improvement: `subject_strict` and `issuer_strict`.

Lastly, probably there also needs to be a parameter that takes in a list of paths to certificate files, drop all of these + the actual certificate in a StoreContext and run https://pyopenssl.org/en/stable/api/crypto.html#OpenSSL.crypto.X509StoreContext.verify_certificate on all of them to be able to check if a certificate has a valid path to a CA certificate.
Another probably upcoming parameter is the `authorityKeyIdentifier` field as well as checks for the criticality of various extensions (see `openssl_csr`)

MarkusTeufelberger commented on test/integration/targets/openssl_certificate/tasks/main.yml:
```diff
@@ -71,4 +71,4 @@
     - import_tasks: ../tests/validate.yml
 
-  when: pyopenssl_version.stdout|version_compare('15.0.0', '>=')
+  when: pyopenssl_version.stdout|version_compare('0.15', '>=')
```
@Spredzy - are you sure about that one? Seems like it fails on shippable: https://app.shippable.com/github/ansible/ansible/runs/36756/summary/console

Spredzy replied:
@MarkusTeufelberger Yep, pretty sure. The issue the CI is reporting is about `openssl_csr` module.

> 2017-09-11 15:53:29 TASK [openssl_csr : Generate CSR]
> 2017-09-11 15:53:30 fatal: [testhost]: FAILED! => {"changed": false, "failed": true, "msg": "You need to have PyOpenSSL>=0.15 to generate CSRs"}

Since 6819e71 is ensuring `pyOpenSSL>=0.15`, the same `block/when` condition should be added in the `openssl_csr/tasks/main.yml` file

Spredzy commented:
@MarkusTeufelberger regarding your last comment about next improvements, (kubernetes, more _strict, more extensions) could you fill a bug for each of them, so we can track them.

I'll happily work on it, I am currently reading [rfc5280](https://www.ietf.org/rfc/rfc5280.txt), to ensure those openssl modules are both compliant and offer as many feature as the rfc mentions.

So yep, any improvement, new ideas is welcome. As for this patch, I'd like it to stay as is - since its a bug fix so it could be merged in stable-2.4

MarkusTeufelberger commented:
Sure, I wanted to file issues for them yesterday already, but @ansibot ran a bit wild in there (migrating all issues from a different repo in here...) so I didn't want to mess with its mojo.

This PR will of course only include integration tests and fixes to issues that were uncovered within these tests (so they pass when it gets merged). It was more of a heads-up on what me and @equinox0815 are still planning to have, since we're building upon these modules internally. Since there are sometimes issues coming up, I tried to trigger them with the integration tests, so they don't affect others and are reproducible. :)

As the bot seems to be done by now, I'll start filing individual issues, so we can also discuss who should do what.

Spredzy commented:
Great, sounds awesome :+1: 

Also for the CI issue, not sure if you've seen my comment.

The issue the CI is reporting is about `openssl_csr` module.

> 2017-09-11 15:53:29 TASK [openssl_csr : Generate CSR]
> 2017-09-11 15:53:30 fatal: [testhost]: FAILED! => {"changed": false, "failed": true, "msg": "You need to have PyOpenSSL>=0.15 to generate CSRs"}

Since commit 6819e71 is ensuring `pyOpenSSL>=0.15`, the same block/when condition should be added in the `openssl_csr/tasks/main.yml` file

Commit:
check pyopenssl version in openssl_csr integration test

MarkusTeufelberger commented:
Done with writing up the issues and this PR probably as well for now.

shipit

resmo commented:
LGTM
shipit

Spredzy commented:
shipit

Spredzy commented:
bot_status

abadger commented:
merged to devel and cherry-picked to temp-staging-post-2.4.0 for the 2.4.1 release.
